### PR TITLE
shell_commands: adapt ping6 to original-ping-like implementation

### LINKED ISSUE #9387: shell `ping6` doesn't display duplicate packages
#### Description
This is rather a question. Maybe there is a good reason for this behavior.

When pinging and duplicates arrive, the riot shell `ping6` command just says
```
12 bytes from fe80::xxx: id=84 seq=1 hop limit=64 time = 0.280 ms
dropping additional response packet (probably caused by duplicates)
```

IPutils ping does it like this:
```
64 bytes from fe80::yyy%wlp3s0: icmp_seq=1 ttl=64 time=0.051 ms
64 bytes from fe80::zzz%wlp3s0: icmp_seq=1 ttl=64 time=166 ms (DUP!)
```

I wonder why you chose to do it like this. My primary reason to ask is that I use `ping ff02::1` etc for debugging the network, for which the latter output is way more useable. Also, in general, when you have duplicates, I suppose you'd want to know where they are from. (Plus, it is quite arbitrary which of the packets are "hidden" by not displaying their information.)

#### Steps to reproduce the issue
Set up a node in a way that there's at least two other hosts listening on the same network.
Run `ping6 ff02::1`

#### Versions
linux
             native gcc: gcc (GCC) 8.1.1 20180531
      arm-none-eabi-gcc: arm-none-eabi-gcc (Arch Repository) 8.1.0

### PR DESCRIPTION
**Pull request by miri64**
### Contribution description
Loosely based on [the original ping] and [netutil]'s ping

New features (compared to old RIOT version):
 - Non-positional parameters
 - Configurable hop-limit
 - Better duplicate detection (addresses #9387)
 - Better asynchronous behavior
 - Potential for future move to `sock_ip`
 - (Optional) DNS-support
 - Multithreading-safe (in case shell-command handler gets called
   from multiple threads)

[the original ping]: http://ftp.arl.army.mil/~mike/ping.html
[netutil]: https://www.gnu.org/software/inetutils/

### Issues/PRs references
Resolves #9387.

### CODE REVIEW

**Comment by miri64** [comment]
(I originally looked at the busybox-implementation [that's where the original title of this PR came from, because the commit had still the name ^^"], but since that is licensed under 4-clause BSD and thus incompatible to (L)GPL, I needed to dig deeper and was able to reconstruct my own version from the original ping's source [public domain; can be found on wayback machine] and inetutil's [licensed under GPLv3] and thus my version is now rather inspired by those works than the busybox one)

**Comment by miri64** [comment]
For the record: this implementation is also more efficient than the old one. On a Cortex-M4 platform it uses ~200 byte less ROM and 56 byte less RAM than the previous implementation.

**Comment by miri64** [comment]
(picked `nucleo-f446ze` to check that)

**Comment by jcarrano** [comment]
I'm running it without arguments and it's spitting a stream of random bytes to the terminal.

**Comment by miri64** [comment]
> I'm running it without arguments and it's spitting a stream of random bytes to the terminal.

On which application and which board?

**Comment by miri64** [comment]
mh curious, yesterday it was working...

**Comment by miri64** [comment]
@jcarrano can you try again please?

**Comment by jcarrano** [comment]
@miri64 Yes, it's fixed. I could run it. I can't test hoplimit because my test setup is quite simple. I'll start looking at the code.

File: sys/shell/commands/sc_gnrc_icmpv6_echo.c [hunk 1]
```diff
@@ -0,0 +1,432 @@
...
+            /* requeue other packets */
+            msg_send(&msg, sched_active_pid);
+        }
+    }
+    return res;
+}
+
+static void _usage(char *cmdname)
+{
+    printf("%s [-c <count>] [-h] [-i <ms interval>] [-s <packetsize>]\n",
```
Comment (jcarrano) [inline sys/shell/commands/sc_gnrc_icmpv6_echo.c]: If the idea is to save flash bytes, then this whole function can be written with two calls to `fputs`.

File: sys/shell/commands/sc_gnrc_icmpv6_echo.c [hunk 2]
```diff
@@ -0,0 +1,432 @@
...
+    int res = 0;
+
+    /* jump over cmdname */
+    if ((--argc) == 0) {
+        /* only command name given */
+        res = 1;
+    }
+    ++argv;
+    /* parse command line arguments */
+    for (; argc > 0; argc--, argv++) {
```
Comment (jcarrano) [inline sys/shell/commands/sc_gnrc_icmpv6_echo.c]: I find this simultaneous decrementing and incrementing kind of fragile. Why not use an explicit index?

File: sys/shell/commands/sc_gnrc_icmpv6_echo.c [hunk 3]
```diff
@@ -0,0 +1,432 @@
...
+    puts("     packetsize: number of bytes in echo payload; must be >= 4 to "
+              "measure round trip time (default: 4)");
+    puts("     hoplimit: Set the IP time to life/hoplimit "
+              "(default: interface config)");
+    puts("     ms timeout: Time to wait for a resonse in milliseconds "
+              "(default: 1000). he option affects only timeout in absence "
+              "of any responses, otherwise wait for two RTTs");
+}
+
+static int _configure(int argc, char **argv, _ping_data_t *data)
```
Comment (jcarrano) [inline sys/shell/commands/sc_gnrc_icmpv6_echo.c]: `ping6 -c 3` without hostname fails. I have some quite lightweight code for parsing command line arguments that I wrote several years ago. If there's interest I can bring it back to life. I think there's no point in everyone hacking together their own parser.

File: sys/shell/commands/sc_gnrc_icmpv6_echo.c [hunk 4]
```diff
@@ -0,0 +1,432 @@
...
+{
+    char *cmdname = argv[0];
+    int res = 0;
```
Comment (jcarrano) [inline sys/shell/commands/sc_gnrc_icmpv6_echo.c]: If there are required arguments, then the default state should be "failure" until those arguments are found. From what I see, there is no checking that the hostname is given.

File: sys/shell/commands/sc_gnrc_icmpv6_echo.c [hunk 5]
```diff
@@ -0,0 +1,432 @@
...
+                    res = 1;
+                    break;
+            }
+        }
+    }
+    if (res != 0) {
+        _usage(cmdname);
+    }
+#ifdef MODULE_LUID
+    luid_custom(&data->id, sizeof(data->id), DEFAULT_ID);
```
Comment (jcarrano) [inline sys/shell/commands/sc_gnrc_icmpv6_echo.c]: This gets done even if the command gets called without arguments. Is that a problem?

**Comment by miri64** [comment]
> @miri64 Yes, it's fixed. I could run it. I can't test hoplimit because my test setup is quite simple. I'll start looking at the code.

https://www.iot-lab.info/ ;-)

File: sys/shell/commands/sc_gnrc_icmpv6_echo.c (hunk 1 above)
Comment (miri64) [reply sys/shell/commands/sc_gnrc_icmpv6_echo.c]: `fputs` is not portable to our current MSP430 port:

```
$ BOARD=z1 make -C examples/gnrc_networking
[…]
/home/mlenders/Repositories/RIOT-OS/RIOT/sys/shell/commands/sc_gnrc_icmpv6_echo.c: In function ‘_usage’:
/home/mlenders/Repositories/RIOT-OS/RIOT/sys/shell/commands/sc_gnrc_icmpv6_echo.c:150:5: error: implicit declaration of function ‘fputs’ [-Werror=implicit-function-declaration]
/home/mlenders/Repositories/RIOT-OS/RIOT/sys/shell/commands/sc_gnrc_icmpv6_echo.c:150:11: error: ‘stdout’ undeclared (first use in this function)
/home/mlenders/Repositories/RIOT-OS/RIOT/sys/shell/commands/sc_gnrc_icmpv6_echo.c:150:11: note: each undeclared identifier is reported only once for each function it appears in
```

File: sys/shell/commands/sc_gnrc_icmpv6_echo.c (hunk 5 above)
Comment (miri64) [reply sys/shell/commands/sc_gnrc_icmpv6_echo.c]: `data` is a on the stack of the calling thread and get's re-initialized every time the shell handler is called, so no harm done (the counter in luid may go up, but so does e.g. the PID counter in UNIX as well).

**Comment by jcarrano** [comment]
@miri64 
> Your registration has been saved and submitted to validation by an administrator.

File: sys/shell/commands/sc_gnrc_icmpv6_echo.c (hunk 3 above)
Comment (miri64) [reply sys/shell/commands/sc_gnrc_icmpv6_echo.c]: Fixed.

File: sys/shell/commands/sc_gnrc_icmpv6_echo.c (hunk 2 above)
Comment (miri64) [reply sys/shell/commands/sc_gnrc_icmpv6_echo.c]: For `argc`/`argv` I find this usually ok, but let me try with implicit index.

File: sys/shell/commands/sc_gnrc_icmpv6_echo.c (hunk 4 above)
Comment (miri64) [reply sys/shell/commands/sc_gnrc_icmpv6_echo.c]: Fixed

File: sys/shell/commands/sc_gnrc_icmpv6_echo.c (hunk 3 above)
Comment (miri64) [reply sys/shell/commands/sc_gnrc_icmpv6_echo.c]: > I have some quite lightweight code for parsing command line arguments that I wrote several years ago. If there's interest I can bring it back to life. I think there's no point in everyone hacking together their own parser.

While I agree, but that opens a similar can of worms as the POSIX networking part, as some libc-variants already provide `getopt` and some don't. Since the `shell` is purely a debugging feature (and shell commands should also have as less dependencies as possible), I'd rather prefer every shell command to handle parsing for itself.

File: sys/shell/commands/sc_gnrc_icmpv6_echo.c (hunk 1 above)
Comment (jcarrano) [reply sys/shell/commands/sc_gnrc_icmpv6_echo.c]: What a pain! At least it should be possible do only one call to `puts()` by pasting all the lines together.

File: sys/shell/commands/sc_gnrc_icmpv6_echo.c (hunk 2 above)
Comment (miri64) [reply sys/shell/commands/sc_gnrc_icmpv6_echo.c]: Ok, going for explicit index.

File: sys/shell/commands/sc_gnrc_icmpv6_echo.c (hunk 3 above)
Comment (jcarrano) [reply sys/shell/commands/sc_gnrc_icmpv6_echo.c]: getopt posix-only, and does not support long options (getopt_long is a GNU extension).

> I'd rather prefer every shell command to handle parsing for itself.

I would rephrase that as "I'd prefer every shell command to have it's own parsing bugs".

File: sys/shell/commands/sc_gnrc_icmpv6_echo.c (hunk 1 above)
Comment (miri64) [reply sys/shell/commands/sc_gnrc_icmpv6_echo.c]: Then `cmdname` needs to be constant which it isn't (or we have a weird line-break).

File: sys/shell/commands/sc_gnrc_icmpv6_echo.c (hunk 3 above)
Comment (miri64) [reply sys/shell/commands/sc_gnrc_icmpv6_echo.c]: > getopt posix-only, and does not support long options (getopt_long is a GNU extension).

At least my version of `newlib` supports it:

```
$ grep -R getopt /usr/arm-none-eabi/include/
/usr/arm-none-eabi/include/getopt.h:getopt.h - Read command line options
/usr/arm-none-eabi/include/getopt.h:The getopt() function parses the command line arguments.  Its arguments argc
/usr/arm-none-eabi/include/getopt.h:The getopt_long() function works like getopt() except that it also accepts
/usr/arm-none-eabi/include/getopt.h:The getopt() function returns the option character if the option was found
/usr/arm-none-eabi/include/getopt.h:The getopt_long() function's return value is described below.
/usr/arm-none-eabi/include/getopt.h:The function getopt_long_only() is identical to getopt_long(), except that a
/usr/arm-none-eabi/include/getopt.h:getopt() and friends to return EOF with optind != argc.
/usr/arm-none-eabi/include/getopt.h:This file and the accompanying getopt.c implementation file are hereby 
/usr/arm-none-eabi/include/getopt.h:    int *flag;			/* determines if getopt_long() returns a
/usr/arm-none-eabi/include/getopt.h:/* While getopt.h is a glibc extension, the following are newlib extensions.
/usr/arm-none-eabi/include/getopt.h: * They are optionally included via the __need_getopt_newlib flag.  */
/usr/arm-none-eabi/include/getopt.h:#ifdef __need_getopt_newlib
/usr/arm-none-eabi/include/getopt.h:     allocated variable of type struct getopt_data.  */
/usr/arm-none-eabi/include/getopt.h:  #define getopt_r		__getopt_r
/usr/arm-none-eabi/include/getopt.h:  #define getopt_long_r		__getopt_long_r
/usr/arm-none-eabi/include/getopt.h:  #define getopt_long_only_r	__getopt_long_only_r
/usr/arm-none-eabi/include/getopt.h:  /* The getopt_data structure is for reentrancy. Its members are similar to
/usr/arm-none-eabi/include/getopt.h:  typedef struct getopt_data
/usr/arm-none-eabi/include/getopt.h:  } getopt_data;
/usr/arm-none-eabi/include/getopt.h:#endif /* __need_getopt_newlib */
/usr/arm-none-eabi/include/getopt.h:  int _EXFUN (getopt,
/usr/arm-none-eabi/include/getopt.h:  int _EXFUN (getopt_long,
/usr/arm-none-eabi/include/getopt.h:  int _EXFUN (getopt_long_only,
/usr/arm-none-eabi/include/getopt.h:#ifdef __need_getopt_newlib
/usr/arm-none-eabi/include/getopt.h:  int _EXFUN (__getopt_r,
/usr/arm-none-eabi/include/getopt.h:	       struct getopt_data * __data));
/usr/arm-none-eabi/include/getopt.h:  int _EXFUN (__getopt_long_r,
/usr/arm-none-eabi/include/getopt.h:	       struct getopt_data * __data));
/usr/arm-none-eabi/include/getopt.h:  int _EXFUN (__getopt_long_only_r,
/usr/arm-none-eabi/include/getopt.h:	       struct getopt_data * __data));
/usr/arm-none-eabi/include/getopt.h:#endif /* __need_getopt_newlib */
/usr/arm-none-eabi/include/getopt.h:/* END OF FILE getopt.h */
/usr/arm-none-eabi/include/sys/unistd.h:# include <getopt.h>
/usr/arm-none-eabi/include/sys/unistd.h:extern char *optarg;			/* getopt(3) external variables */
/usr/arm-none-eabi/include/sys/unistd.h:int	 getopt(int, char * const [], const char *);
/usr/arm-none-eabi/include/sys/unistd.h:extern int optreset;			/* getopt(3) external variable */
```

Compare e.g. AVRlibc:

```
$ grep -R getopt /usr/lib/avr/include/
```

**Comment by tcschmidt** [comment]
@cgundogan maybe you can test this on IoTLab?

**Comment by cgundogan** [comment]
>@miri64
>
>>Your registration has been saved and submitted to validation by an administrator.

@jcarrano were you able to perform the tests in a multihop setup? If not, I'd help you out here.

**Comment by miri64** [comment]
ping @jcarrano

**Comment by jcarrano** [comment]
@miri64 I was not able to perform the test. I would have to learn how to use IOTLab and I'm lacking time ATM.

**Comment by miri64** [comment]
@jcarrano you can also use two native nodes or two of the many IEEE 802.15.4 capable nodes on your desk ;-)

**Comment by miri64** [comment]
How do we proceed here?

**Comment by jcarrano** [comment]
@miri64 Sorry for leaving this PR hanging. I have since given up reviewing network-related stuff which, not being in my main area of expertise, takes me too much time while I have little to offer in terms of a valuable review.

My own reviews in this PR were mainly related to command line parsing, and they have been addressed, so I will dismiss myself if you don't mind.

**Comment by miri64** [comment]
> […] so I will dismiss myself if you don't mind.

Sure. I understand.

**Comment by gschorcht** [comment]
@miri64 I have tested the `ping6` command. It is working as expected. What about option `-I` to define the interface as implemented on UNIX machines? It is doing the same as as the `%`- notation.

**Comment by miri64** [comment]
> @miri64 I have tested the `ping6` command. It is working as expected. What about option `-I` to define the interface as implemented on UNIX machines? It is doing the same as as the `%`- notation.

The [man page to `iputils` ping](https://linux.die.net/man/8/ping) says:

>     -I interface address
>         Set source address to specified interface address. Argument may be numeric IP
>         address or name of device. When pinging IPv6 link-local address this option is
>         required.

So I would say it's redundant in a pure IPv6 scenario.

**Comment by gschorcht** [comment]
> > ```
> > -I interface address
> >     Set source address to specified interface address. Argument may be numeric IP
> >     address or name of device. When pinging IPv6 link-local address this option is
> >     required.
> > ```
> So I would say it's redundant in a pure IPv6 scenario.

Yes, was just a question whether we should have it in addition to the `%`-notation of interface id.

**Comment by miri64** [comment]
> Yes, was just a question whether we should have it in addition to the `%`-notation of interface id.

I'd say no. One way to define the interface is enough IMHO (contrary to iputil's ping you can also use the `%` with a global address)

File: sys/shell/commands/sc_gnrc_icmpv6_echo.c [hunk 6]
```diff
@@ -0,0 +1,425 @@
...
+                break;
+            }
+            case _SEND_NEXT_PING:
+                _pinger(&data);
+                break;
+            case _PING_FINISH:
+                goto finish;
+            default:
+                /* requeue wrong packets */
+                msg_send(&msg, sched_active_pid);
```
Comment (cgundogan) [inline sys/shell/commands/sc_gnrc_icmpv6_echo.c]: What other messages can appear here? Shouldn't they be dropped instead of requeued? I can't see another `msg_recv` anywhere here that handles other types of messages, so the queue would get full with garbage ..

Comment (miri64) [reply sys/shell/commands/sc_gnrc_icmpv6_echo.c]: I can research that, but on first glance I would say: remember that this is the main thread. Other application specific messages might land in the queue.

Comment (miri64) [reply sys/shell/commands/sc_gnrc_icmpv6_echo.c]: (I did it because the current version in master does it as well: https://github.com/RIOT-OS/RIOT/blob/1d693403b65231b6784a25f6bb5a1c2657af6834/sys/shell/commands/sc_icmpv6_echo.c#L302-L305)

Comment (cgundogan) [reply sys/shell/commands/sc_gnrc_icmpv6_echo.c]: >remember that this is the main thread. Other application specific messages might land in the queue.

yes, sounds sensible. Please research that anyway (:

Comment (miri64) [reply sys/shell/commands/sc_gnrc_icmpv6_echo.c]: Ok, I had in memory, that this line existed because someone asked for it, but apparently I introduced it myself in https://github.com/RIOT-OS/RIOT/pull/2555. Anyway, given that this is `main` and that we didn't face any problems with this until now, I think the line is here correctly.

**Comment by cgundogan** [comment]
squashing needed

**Comment by miri64** [comment]
> We should also scan the code base and tutorials/wiki to adjust for the new syntax. e.g. [here](https://github.com/RIOT-OS/RIOT/wiki/How-to-install-6LoWPAN-Linux-Kernel-on-Raspberry-Pi#testing-from-riot).

We should this however do in a follow-up IMHO

**Comment by miri64** [comment]
Squashed

**Comment by cgundogan** [comment]
Alright. Murdock also agrees!

**Comment by miri64** [comment]
> Tested on native and iotlab-m3 with unicast and multicast addresses. The output looks sensible. ACK.
> We should also scan the code base and tutorials/wiki to adjust for the new syntax. e.g. [here](https://github.com/RIOT-OS/RIOT/wiki/How-to-install-6LoWPAN-Linux-Kernel-on-Raspberry-Pi#testing-from-riot).

@cgundogan the wiki page you referenced is already marked as outdated. Should we really fix it there? (all other occurrences I found, I fixed).

**Comment by miri64** [comment]
(it even mentions `fibroute`)
